pkgs/development/compilers/llvm: use makeScopeWithSplicing

LunNova opened the pull request:
Rebase of @RossComputerGuy's PR #436350 + ROCm scope usage & rebuild avoidance + darwin rebuild avoidance.

## Things done

- [x] Ensure no rebuilds
- [x] why is `.override` trying to disappear from darwin default llvmPackages?
  - Because override and overrideScope don't compose without [changes to makeOverridable](https://github.com/NixOS/nixpkgs/pull/445668#issuecomment-3328834974)
  - We need these to compose for cases like https://github.com/NixOS/nixpkgs/blob/2458def781a8a3e34d76b3561adc2f9bcef2a3b7/pkgs/build-support/build-mozilla-mach/default.nix#L207-L214
  - To get to zero rebuilds we need to propagate an incorrect `override` function that discards all of the stdenv overrides, because that's what happened in the past
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

RossComputerGuy committed:
pkgs/development/compilers/llvm: use makeScopeWithSplicing'

Co-authored-by: Luna Nova <git@lunnova.dev>
Co-authored-by: Emily <vcs@emily.moe>

LunNova committed:
rocmPackages.llvm: use llvmPackages.overrideScope

Uses new overrideScope feature to override base LLVM packages
without risking leaking incorrect non-overridden packages
from the base LLVM version.

To achieve no rebuilds or new attrs:
- Explicit list of packages to expose in an inherit instead
  of the entire scope
- Noting where workarounds for lack of overrideScope should
  be removed in a followup such as re-attaching a
  replaceVars patch or disabling separate tblgen package,
  but are kept for now to avoid a rebuild

To avoid a large formatting diff a small helper has
been introduced that allows an overrideScope call with a single
arg, convincing nixfmt not to indent the scope body further.

LunNova commented:
Ran into a problem. darwin's `llvmPackages` ends up not having a `.override`, causing an eval failure if that's ever used. darwin applies an `overrideScope` in `pkgs/stdenv/darwin/default.nix` which ends up removing that attr.

This happens on this rebased version, it also was happening on the original. No eval error must be because in that version nothing was calling .override on the default version of `llvmPackages`.

```shell
# PR 436350's override missing for darwin
$ nix eval github:nixos/nixpkgs/acd183d2efb23f3877c448feec7d76bf6125b5dd#legacyPackages.aarch64-darwin.llvmPackages.override
# but it's present for the other versions
$ nix eval github:nixos/nixpkgs/acd183d2efb23f3877c448feec7d76bf6125b5dd#legacyPackages.aarch64-darwin.llvmPackages_20.override
{ __functionArgs = { bootBintools = true; bootBintoolsNoLibc = true; buildPackages = false; callPackage = false; generateSplicesForMkScope = false; lib = false; llvmVersions = true; patchesFn = true; pkgs = false; recurseIntoAttrs = false; stdenv = false; stdenvAdapters = false; targetPackages = false; }; __functor = «lambda __functor @ /nix/store/ny1zr85wfp3yishpwfpqhh3dpnslrl6g-source/lib/trivial.nix:989:17»; }
```

For 0 rebuilds we probably need to make darwin stdenv's `llvmPackages` immediately `// { inherit (super.llvmPackages) override; }` since that's what it ends up as in `master`, but that wouldn't be a good permanent fix.

There must be some mechanism that allows this to work for `(something.overrideAttrs{}).override{}`, which I'll try to find. cc @ShamrockLee since I vaguely recall you touching related override machinery for a similar problem when you were teaching python packages to compose overrides well.

LunNova commented:
Linux: all good
darwin: oops stdenv rebuild

```diff
- /nix/store/s8a14scnv3pryics9c88i50q92ybilvq-clang-wrapper-21.1.1.drv:{out}
+ /nix/store/als7gpcr1iszsvpmh5fjx6zb7nibbikb-clang-wrapper-21.1.1.drv:{out}
• The builders do not match
    - /nix/store/xhsnsmvh1ka5mxszd1psxq36vaanb8jy-bash-5.3p3/bin/bash
    + /nix/store/linnqbx1p7f1wfc4qi96kcifnwa0b97s-bash-5.3p3/bin/bash
• The input derivation named `apple-sdk-11.3` differs
  - /nix/store/6823hji7rfnjk1188ys6fzsrk6fj6vpk-apple-sdk-11.3.drv:{out}
  + /nix/store/p4vcq7sya799apnnxnafk34z3dwm60xm-apple-sdk-11.3.drv:{out}
  • The builders do not match
      - /nix/store/xhsnsmvh1ka5mxszd1psxq36vaanb8jy-bash-5.3p3/bin/bash
      + /nix/store/linnqbx1p7f1wfc4qi96kcifnwa0b97s-bash-5.3p3/bin/bash
  • The input derivation named `CoreSymbolication-0-unstable-2018-06-17` differs
    - /nix/store/5bngcwc1mgqvr32l3ngf6rpzqfi0bhgz-CoreSymbolication-0-unstable-2018-06-17.drv:{out}
    + /nix/store/1c608hy0ivpi63jgyisv4apprw4s8fqp-CoreSymbolication-0-unstable-2018-06-17.drv:{out}
    • The builders do not match
        - /nix/store/xhsnsmvh1ka5mxszd1psxq36vaanb8jy-bash-5.3p3/bin/bash
        + /nix/store/linnqbx1p7f1wfc4qi96kcifnwa0b97s-bash-5.3p3/bin/bash
    • The input derivation named `bash-5.3p3` differs
      - /nix/store/qmsjcc3pklj88mx6ikg9l1yn8v317597-bash-5.3p3.drv:{out}
      + /nix/store/6ym0a08dj88s8j6rlll1waqv0xrd1m0w-bash-5.3p3.drv:{out}
      • The input derivation named `bootstrap-stage2-clang-wrapper-21.1.1` differs
        - /nix/store/w93giv1a2q81ydwlmygzswhb3afwhzgq-bootstrap-stage2-clang-wrapper-21.1.1.drv:{out}
        + /nix/store/423ddx7a6ipq3wx3a6i1fv7wc10lc1fv-bootstrap-stage2-clang-wrapper-21.1.1.drv:{out}
        • The input derivation named `compiler-rt-libc-21.1.1` differs
          - /nix/store/96mrz0z9k7n8kgnbkixs50pym1jnlp23-compiler-rt-libc-21.1.1.drv:{dev,out}
          + /nix/store/2g4y0ahb4hg2ahzsbgy6a914n8b9imsh-compiler-rt-libc-21.1.1.drv:{dev,out}
          • The input derivation named `bootstrap-stage-xclang-stdenv-darwin` differs
            - /nix/store/fs7l66x23l84qpzsh5jlvxk4wx53d9zl-bootstrap-stage-xclang-stdenv-darwin.drv:{out}
            + /nix/store/3755q34jwd8099clnjgi72sv68lchyn2-bootstrap-stage-xclang-stdenv-darwin.drv:{out}
            • The input derivation named `clang-wrapper-21.1.1` differs
              - /nix/store/gm4ipb0xrk7r4v0mvnml7wf6q9a3rqy9-clang-wrapper-21.1.1.drv:{out}
              + /nix/store/148wya5xnmcbx6byvfc93fdwzp8vcd6g-clang-wrapper-21.1.1.drv:{out}
              • The set of input derivation names do not match:
                  - libcxx-19.1.2+apple-sdk-15.5
                  + libcxx-21.1.1
              • The environments do not match:
                  depsTargetTargetPropagated=''
                  /nix/store/8f834dcac4msj5wiqipb7j94b2pk1d78-libcxx-19.1.2+apple-sdk-15.5/nix/store/kgg9k4sprb81idr8bc46wblmc9l4hy2i-libcxx-21.1.1-dev /nix/store/73i5cccnr4jw67ksj26lm5p2nxr1x2zx-compiler-rt-21.1.1-dev
              ''
```

LunNova committed:
pkgs/stdenv/darwin: rebuild avoidance for LLVM scope

- Setting recurseForDerivations again post overrideScope
- Hackily propagating initial value llvmPackages.override

emilazy commented on pkgs/development/compilers/llvm/common/default.nix:
```diff
@@ -109,12 +105,23 @@ let
+makeScopeWithSplicing' {
+  inherit otherSplices;
+  f =
+    llvmPackages:
```
```suggestion
  f =
    self:
```
This is the naming convention used elsewhere, and `llvmPackages` is ambiguous – e.g. using `llvmPackages` in `callPackage` arguments inside the scope would not give this scope and would be misleading. Not a hard blocker, but I think this would be less confusing.

emilazy commented on pkgs/development/compilers/llvm/common/default.nix:
```diff
@@ -134,37 +141,41 @@ let
         mkExtraBuildCommands0 cc
         + ''
-          ln -s "${targetLlvmLibraries.compiler-rt-no-libc.out}/lib" "$rsrc/lib"
+          ln -s "${targetLlvmPackages.compiler-rt-no-libc.out}/lib" "$rsrc/lib"
```
This would be
```suggestion
          ln -s "${self.compiler-rt-no-libc.__spliced.targetTarget.out}/lib" "$rsrc/lib"
```

emilazy commented on pkgs/development/compilers/llvm/common/default.nix:
```diff
@@ -383,126 +393,103 @@ let
+          if stdenv.targetPlatform.libc == "llvm" then buildPackages.cmakeMinimal else buildPackages.cmake;
+        python3 =
+          if stdenv.targetPlatform.libc == "llvm" then
+            buildPackages.python3Minimal
```
(This explicit `buildPackages` use is a bit screwy – these are used in `nativeBuildInputs` anyway, so this just breaks splicing for seemingly no reason. However, this is a pre‐existing problem, of course; I’m just noting it here as I pore over the cross stuff here.)

emilazy commented on pkgs/development/rocm-modules/6/llvm/default.nix:
```diff
@@ -258,13 +279,8 @@ let
-  # clang-orig = llvmPackagesRocm.clang; # nix why-depends --derivation .#rocmPackages.clr .#rocmPackages.llvm.clang-orig
-  llvm = llvmPackagesRocm.llvm.overrideAttrs (old: {
+overrideLlvmPackagesRocm (s: {
+  libllvm = (s.prev.libllvm.override { }).overrideAttrs (old: {
```
Remind me what the `.override { }` is doing here? I think I knew once, but then forgot. It should of course be a no‐op, so it scares me a little if it’s not – is it related to the rebuild avoidance? Will it start being a no‐op again after we make a follow‐up change?

emilazy reviewed (commented):
Thanks for picking this up, and for @RossComputerGuy for the original PR. I am very glad to see rebuilds down to zero and that exposing all the ways in which things were screwy before :)

I think there is one important change we should make around how this works with cross, and I’d like to understand what we’re going to do about `llvmPackages.override` in the Darwin `stdenv`, but this looks like a big improvement to me in general and I hope we can merge this very soon.

LunNova commented on pkgs/development/rocm-modules/6/llvm/default.nix:
```diff
@@ -521,4 +526,7 @@ rec {
-}
+  # AMD has a separate MLIR impl which we package under rocmPackages.rocmlir
+  # It would be an error to rely on the original mlir package from this scope
+  mlir = null;
```
Newly introduced, to me it made sense to include with introducing `overrideScope` usage here as it doesn't cause a rebuild. Without overrideScope there wasn't a way to enforce this.

emilazy committed:
llvmPackages.openmp: don’t use a target platform LLVM

This was used in 4b07aeae16811bcb6ab2cef67133180bcb30ed67 to get
what was actually a *host* LLVM in practice, when the `buildHost`
LLVM package set was included in the `callPackage` scope used for
libraries. It should no longer be necessary and is now quite confusing.

emilazy committed:
llvmPackages: remove unnecessary use of `pkgs`

emilazy committed:
llvmPackages: add comments about suspicious things

emilazy committed:
llvmPackages: remove things made unnecessary by splicing

LunNova commented on pkgs/stdenv/darwin/default.nix:
```diff
@@ -466,87 +466,91 @@ assert bootstrapTools.passthru.isFromBootstrapFiles or false; # sanity check
+          ))
+          // {
+            inherit (super.llvmPackages) override;
+          };
```
> although AIUI @Artturin says you should always use callPackage in overlays anyway for splicing reasons, so maybe it would actually be good??

Is there discussion elsewhere about this to follow?

emilazy replied:
I forget where I read @Artturin saying this, I’m afraid. But it’s because of `final.callPackage` having access to the spliced packages but `final` not.

LunNova commented:
Thanks @emilazy, that looks great*!

I can prep followup changes that drop the rebuild avoidance introduced here.

*skipping over some consequences of past decisions that maybe do not look great
